src/core/geometry/path/transform.js

```diff
@@ -0,0 +1,18 @@
+const canonicalize = require('./canonicalize')
+const fromPointArray = require('./fromPointArray')
+const vec3 = require('../../math/vec3')
+
+/**
+ * Produces a path by transforming all points in the provided path.
+ * @param {mat4} matrix - the matrix to transform with.
+ * @param {path2} path - the path to transform.
+ * @returns {path2} - the transformed path.
+ * @example
+ * transform(fromZRotation(degToRad(90)), path)
+ */
+const transform = (matrix, path) =>
+    fromPointArray(
+        { closed: path.isClosed },
+        canonicalize(path).points.map(point => vec3.transform(matrix, point)))
```

Comment:
Canonicalize ensures quantized points for transformation.

Transform produces a non-canonical result.

Reply:
> Canonicalize ensures quantized points for transformation.

I don’t see why ‘quantized’ points are require to perform transforms. Or should I ask... are ‘quantized’ points required before performing operations of any kind?

Reply:
Let's generalize it to "Operations that depend on values require canonical values."
For vec3 at the moment that involves quantization.

If we perform transforms on unquantized values we run the risk of incremental error.

--

Elsewhere I've been moving to toward having separate (canonicalized) base data and separate transformed data.

In that approach you end up with path.points = path.basePoints.map(...) instead of transforming earlier transformed points again.

If you like, I'll move across to that model now -- otherwise I'll follow up on it later.

Reply:
Hmm... well, everything is unquantized by nature due to floating point operations, even the direct calculation of points in the mesh of a sphere. And transform is not really dependent on those values, as the matrix multiplication is pretty straight forward, and applied to everthing. So, again, I ask ... when is canonization required? I'm asking this in a general sense because we don't want to canonalize values before every operation. Or do we?

Reply:
Think of it rather as floating point operations produce non-canonical values.
Using non-canonical values leads to potential incremental drift.
So, ideally, you should canonicalize when you need to use those values for something.

Things that don't require canonicalization of the geometry:
  adding new base values (although the new values should be canonicalized
  transforms

Things that require canonicalization:
  producing output points

I'll move path across to do lazy transforms, and then it should hopefully be clearer.

Reply:
the thing is, yes, there is drift & accumulated error, but in practice (except for some specific tests, and even then , not much) , this was not an issue most of the time and seems like perhaps too much extra operations for no immediate gain ? Is this important , yes, but I would rather solve it after basic stuff is working , and perhaps solve it by switching to Integer based coords as we discussed

Reply:
In practice, it's sufficient to make writing tests a problem.
You will get numbers like 0.99997997973964 which will vary slightly from system to system, and make for flaky tests.
Which is probably why there aren't many of them. :)

For that reason along, it needs to happen now.
If you want to make a better system later, by all means.

And before you argue that you should quantize the tests instead of the actual system ...

Reply:
And, also remember why you have FuzzyFactories ...

Reply:
Don't get me wrong, but the base points being accumulated are 'normalized'. And all the operations are also calling canonicalize(), which creates the exposed points from transformed, and then canonicalize(base point). It seems everything is canonicalized, period.

Thoughts? I think we will do the same for geom3, so let's get this right.

Reply:
Note in particular, that 'transform' does not call canonicalize.
Transform does the complement -- generating a new non-canonical geometry.
This would be true of any lazy operation.

What I'd say is that 'all the things you may legitimately look at are always canonicalized when you look at them'. :)

So we can enumerate those:
  (a) we can see the points we add in, so those are canonicalized.
  (b) we can see the points we get out, so those are canonicalized.
  (c) we can't see the points between (a) and (b), so those don't need to be canonicalized.

So, what I've been working with is a general model like this.
  (d) there is a 'base model' (the points we put in)
  (e) there are transforms, etc, which have no visible output, and these may be accumulated lazily. e.g., transform(x, transform(y, z)) doesn't require any canonicalization.
  (f) canonicalization produces a 'representative model' (the transformed points), which can then be operated on to produce output, and which can be cached. This 'representative model' is produced from the 'base model' where possible, to avoid incremental error accumulation.

Which gives us a situation where, from the outside, the geometry is always canonical.
But from the inside the geometry is only canonical when the outside needs it to be.

Reply:
Umm... I disagree.

You don’t see the points in (a), and the normalize of the base points isn’t necessary.
But, you make sure (b) provides canonicalized points to the world. Done.

Reply:
They were visible to you when you added them, so they should have been canonical at that point.

But I agree that it probably doesn't need to be the geometry's responsibility to ensure that it receives canonical input -- we can probably delegate that to the caller.

Reply:
Good. Then, we can use this same method for geom3, and others.

Reply:
Yes, you should be able to see it in the PRs providing surf2pc and solid3bsp